Clean up xcodebuild's messy output

Our console output when running xCode builds is very noisy and hard to read. We heard very good things about https://github.com/supermarin/xcpretty from another team that runs xCode builds, so would be good to give that a try for Flutter.

With #14586, a failed build's output will be:
```
flutter build ios                                                                                                                                                                                                                 (02-09 00:54:37)
Building flutter tool...
Running "flutter packages get" in t72_base_ios_app...  0.5s
Building com.example.t72BaseIosApp for device (ios-release)...
Found saved certificate choice "iPhone Developer: Xiao Yu". To clear, use "flutter config".
Signing iOS app for device deployment using developer identity: "iPhone Developer: Xiao Yu"
Running Xcode clean...                                0.8s
Running Xcode build...                                5.4s
Failed to build iOS app
Error output from Xcode build:
↳
    ** BUILD FAILED **

Xcode's output:
↳
    === BUILD TARGET Runner OF PROJECT Runner WITH CONFIGURATION Release ===
    Building AOT snapshot in release mode (ios-release)...
    Dart snapshot generator failed with exit code 254
    Error: 'file:///Users/xiao/Development/flutter-test/t72_base_ios_app/lib/main.dart': error: line 15 pos 19: semicolon expected
            })));wrong
                      ^
    Failed to build /Users/xiao/Development/flutter-test/t72_base_ios_app.
    Command /bin/sh failed with exit code 255
```

I think it's good enough.

--- a/packages/flutter_tools/bin/xcode_backend.sh
+++ b/packages/flutter_tools/bin/xcode_backend.sh
@@ -4,7 +4,9 @@
 # found in the LICENSE file.
 
 RunCommand() {
-  echo "♦ $*"
+  if [[ -n "$VERBOSE_SCRIPT_LOGGING" ]]; then
+    echo "♦ $*"
+  fi
   "$@"
   return $?
 }

--- a/packages/flutter_tools/lib/src/ios/mac.dart
+++ b/packages/flutter_tools/lib/src/ios/mac.dart
@@ -277,15 +277,39 @@ Future<XcodeBuildResult> buildXcodeProject({
     );
   }
 
+  final Status cleanStatus =
+      logger.startProgress('Running Xcode clean...', expectSlowOperation: true);
+  final RunResult cleanResult = await runAsync(
+    <String>[
+      '/usr/bin/env',
+      'xcrun',
+      'xcodebuild',
+      'clean',
+    ],
+    workingDirectory: app.appDirectory,
+  );
+  cleanStatus.stop();
+  if (cleanResult.exitCode != 0) {
+    throwToolExit('Xcode failed to clean\n${cleanResult.stderr}');
+  }
+
   final List<String> commands = <String>[
     '/usr/bin/env',
     'xcrun',
     'xcodebuild',
-    'clean',
     'build',
     '-configuration', configuration,
     'ONLY_ACTIVE_ARCH=YES',
   ];
+
+  if (logger.isVerbose) {
+    // An environment variable to be passed to xcode_backend.sh determining
+    // whether to echo back executed commands.
+    commands.add('VERBOSE_SCRIPT_LOGGING=YES');
+  } else {
+    // This will print warnings and errors only.
+    commands.add('-quiet');
+  }
 
   if (developmentTeam != null)
     commands.add('DEVELOPMENT_TEAM=$developmentTeam');
@@ -318,27 +342,28 @@ Future<XcodeBuildResult> buildXcodeProject({
     );
   }
 
-  final Status status = logger.startProgress('Running Xcode build...', expectSlowOperation: true);
-  final RunResult result = await runAsync(
+  final Status buildStatus =
+      logger.startProgress('Running Xcode build...', expectSlowOperation: true);
+  final RunResult buildResult = await runAsync(
     commands,
     workingDirectory: app.appDirectory,
     allowReentrantFlutter: true
   );
-  status.stop();
-  if (result.exitCode != 0) {
+  buildStatus.stop();
+  if (buildResult.exitCode != 0) {
     printStatus('Failed to build iOS app');
-    if (result.stderr.isNotEmpty) {
+    if (buildResult.stderr.isNotEmpty) {
       printStatus('Error output from Xcode build:\n↳');
-      printStatus(result.stderr, indent: 4);
+      printStatus(buildResult.stderr, indent: 4);
     }
-    if (result.stdout.isNotEmpty) {
+    if (buildResult.stdout.isNotEmpty) {
       printStatus('Xcode\'s output:\n↳');
-      printStatus(result.stdout, indent: 4);
+      printStatus(buildResult.stdout, indent: 4);
     }
     return new XcodeBuildResult(
       success: false,
-      stdout: result.stdout,
-      stderr: result.stderr,
+      stdout: buildResult.stdout,
+      stderr: buildResult.stderr,
       xcodeBuildExecution: new XcodeBuildExecution(
         commands,
         app.appDirectory,
@@ -348,7 +373,7 @@ Future<XcodeBuildResult> buildXcodeProject({
   } else {
     // Look for 'clean build/<configuration>-<sdk>/Runner.app'.
     final RegExp regexp = new RegExp(r' clean (.*\.app)$', multiLine: true);
-    final Match match = regexp.firstMatch(result.stdout);
+    final Match match = regexp.firstMatch(buildResult.stdout);
     String outputDir;
     if (match != null) {
       final String actualOutputDir = match.group(1).replaceAll('\\ ', ' ');

--- a/packages/flutter_tools/lib/src/ios/mac.dart
+++ b/packages/flutter_tools/lib/src/ios/mac.dart
@@ -285,6 +285,7 @@ Future<XcodeBuildResult> buildXcodeProject({
       'xcrun',
       'xcodebuild',
       'clean',
+      '-configuration', configuration,
     ],
     workingDirectory: app.appDirectory,
   );

